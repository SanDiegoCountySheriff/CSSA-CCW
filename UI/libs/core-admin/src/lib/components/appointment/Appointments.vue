<!-- eslint-disable vue/multiline-html-element-content-newline -->
<!-- eslint-disable vue-a11y/form-has-label -->
<!-- eslint-disable @intlify/vue-i18n/no-raw-text -->
<!-- eslint-disable vue/valid-v-slot -->
<!-- eslint-disable vue-a11y/no-autofocus -->
<template>
  <v-container fluid>
    <v-data-table
      :headers="state.headers"
      :items="data"
      :search="state.search"
      :loading="isLoading && !isError"
      :loading-text="$t('Loading appointment schedules...')"
      :single-expand="state.singleExpand"
      :expanded.sync="state.expanded"
      :items-per-page="14"
      :footer-props="{
        showCurrentPage: true,
        showFirstLastPage: true,
        firstIcon: 'mdi-skip-backward',
        lastIcon: 'mdi-skip-forward',
        prevIcon: 'mdi-skip-previous',
        nextIcon: 'mdi-skip-next',
      }"
      item-key="id"
      item-class="rowClass"
    >
      <template #top>
        <v-toolbar flat>
          <v-toolbar-title
            class="text-no-wrap pr-4"
            style="text-overflow: clip"
            >{{ $t('Appointments') }}</v-toolbar-title
          >
          <v-spacer></v-spacer>
          <v-container>
            <v-row justify="end">
              <v-col align="right">
                <v-btn
                  :to="Routes.APPOINTMENT_MANAGEMENT_ROUTE_PATH"
                  color="primary"
                  >Appointment Management</v-btn
                >
              </v-col>
              <v-col md="6">
                <v-text-field
                  v-model="state.search"
                  prepend-icon="mdi-filter"
                  label="Filter"
                  placeholder="Start typing to filter"
                  single-line
                  hide-details
                >
                </v-text-field>
              </v-col>
              <!-- <v-col
                md="4"
                class="mr-1"
              >
                <v-btn
                  color="primary"
                  :loading="state.isSelecting"
                  :rounded="$vuetify.breakpoint.mdAndDown"
                  @click="handleFileImport"
                  :fab="$vuetify.breakpoint.mdAndDown"
                  :raised="$vuetify.breakpoint.mdAndDown"
                  :x-small="$vuetify.breakpoint.mdAndDown"
                  elevation="2"
                >
                  <v-icon v-if="$vuetify.breakpoint.mdAndDown">
                    mdi-upload
                  </v-icon>
                  <div v-else>Upload New Appointments</div>
                </v-btn>

                <input
                  ref="uploader"
                  label="Upload New Appointments"
                  class="d-none"
                  type="file"
                  @change="onFileChanged"
                  @click="onInputClick"
                  @keydown="onInputClick"
                  accept=".csv, .xls, .xlsx"
                />
              </v-col>
              <v-col md="1">
                <v-tooltip top>
                  <template #activator="{ on, attrs }">
                    <v-btn
                      elevation="2"
                      fab
                      raised
                      rounded
                      color="primary"
                      x-small
                      v-bind="attrs"
                      v-on="on"
                      href="/documents/ccw-appointment-schedule-template.csv"
                      target="_blank"
                      aria-label="Download the template file here"
                    >
                      <v-icon> mdi-download </v-icon>
                    </v-btn>
                  </template>
                  <span>Download the template file here</span>
                </v-tooltip>
              </v-col> -->
            </v-row>
          </v-container>
        </v-toolbar>
      </template>
      <template #item.status="props">
        <v-chip
          v-if="props.item.status.length !== 0"
          :color="$vuetify.theme.dark ? '' : getColor(props.item.status)"
          :text-color="
            $vuetify.theme.dark ? '' : getTextColor(props.item.status)
          "
          class="ma-0 font-weight-regular"
          small
        >
          {{ props.item.status === 'true' ? 'Scheduled' : 'Not Scheduled' }}
        </v-chip>
        <v-icon
          :color="$vuetify.theme.dark ? '' : 'error'"
          medium
          v-else
        >
          mdi-alert-octagon
        </v-icon>
      </template>
      <template #item.name="props">
        <v-avatar
          color="primary"
          size="30"
          class="mr-1"
        >
          <span class="white--text .text-xs-caption"> {{ $t('JS') }}</span>
        </v-avatar>
        {{ props.item.name }}
      </template>
      <template #item.permit="props">
        <router-link
          :to="{
            name: 'PermitDetail',
            params: { orderId: props.item.permit },
          }"
          style="text-decoration: underline; color: inherit"
        >
          {{ props.item.permit }}
        </router-link>
      </template>
      <template #item.payment="props">
        <v-chip
          v-if="props.item.payment.length !== 0"
          color="primary"
          small
        >
          {{ props.item.payment }}
        </v-chip>
        <v-icon
          color="error"
          medium
          v-else
        >
          mdi-alert-octagon
        </v-icon>
      </template>
      <template #item.deletion="props">
        <v-tooltip bottom>
          <template #activator="{ on, attrs }">
            <AppointmentDeleteDialog
              :appointment="props.item"
              :refetch="appointmentRefetch"
              v-bind="attrs"
              v-on="on"
            />
          </template>
          {{ $t('Remove applicant from appointment') }}
        </v-tooltip>
      </template>
      <template #expanded-item="{ item }">
        <td colspan="2">
          {{ $t(`More info about ${item.name}`) }}
        </td>
      </template>
    </v-data-table>
    <v-snackbar
      v-model="state.snackbar"
      :multi-line="state.multiLine"
    >
      {{ state.text }}

      <template #action="{ attrs }">
        <v-btn
          color="warning"
          text
          v-bind="attrs"
          @click="state.snackbar = false"
        >
          Close
        </v-btn>
      </template>
    </v-snackbar>
  </v-container>
</template>

<script setup lang="ts">
import AppointmentDeleteDialog from '../dialogs/AppointmentDeleteDialog.vue'
import Routes from '@core-admin/router/routes'
import { useAppointmentsStore } from '@shared-ui/stores/appointmentsStore'
import { useQuery } from '@tanstack/vue-query'
import { reactive, ref } from 'vue'

const appointmentsStore = useAppointmentsStore()
const {
  isLoading,
  isError,
  data,
  refetch: appointmentRefetch,
} = useQuery(['appointments'], appointmentsStore.getAppointmentsApi)

const allowedExtension = ['.xls', '.xlsx', '.csv']

const uploader = ref()

const state = reactive({
  isSelecting: false,
  search: '',
  singleExpand: true,
  expanded: [],
  snack: false,
  snackColor: '',
  snackText: '',
  pagination: {},
  headers: [
    {
      text: 'STATUS',
      align: 'start',
      sortable: false,
      value: 'status',
    },
    { text: 'APPLICANT NAME', value: 'name' },
    { text: 'ORDER ID', value: 'permit' },
    { text: 'PAYMENT', value: 'payment' },
    { text: 'DATE', value: 'date' },
    { text: 'TIME', value: 'time' },
    { text: '', value: 'deletion' },
  ],
  multiLine: false,
  snackbar: false,
  text: `Invalid file type provided.`,
})

function handleFileImport() {
  state.isSelecting = true

  // After obtaining the focus when closing the FilePicker, return the button state to normal
  window.addEventListener('focus', () => {
    state.isSelecting = false
  })

  uploader.value.click()
}

function onInputClick(e) {
  e.target.value = ''
}

function onFileChanged(e) {
  if (allowedExtension.some(ext => e.target.files[0].name.endsWith(ext))) {
    appointmentsStore.newAppointmentsFile = e.target.files[0]
    appointmentsStore
      .uploadAppointmentsApi()
      .then(() => {
        state.text = 'Successfully uploaded file.'
        state.snackbar = true
      })
      .catch(() => {
        state.text = 'An API error occurred.'
        state.snackbar = true
      })
  } else {
    state.text = 'Invalid file type provided.'
    state.snackbar = true
  }
}

function getColor(name) {
  if (name === 'New' || name.match(/^\d/)) return '#eff8ff'

  if (name === 'cash' || name === 'card') return '#ecfdf3'

  if (name === 'Set') return '#fffaeb'

  return '#f2f4f7'
}

function getTextColor(name) {
  if (name === 'New' || name.match(/^\d/)) return '#2e90fa'

  if (name === 'cash' || name === 'card') return '#12B76A'

  if (name === 'Set') return '#F79009'

  return '#667085'
}
</script>
