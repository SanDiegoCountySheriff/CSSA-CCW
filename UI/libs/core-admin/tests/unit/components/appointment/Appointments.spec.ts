/* eslint-disable @typescript-eslint/ban-ts-comment */
import Appointments from '@core-admin/components/appointment/Appointments.vue'
import Vuetify from 'vuetify'
import { createTestingPinia } from '@pinia/testing'
import { createLocalVue, mount } from '@vue/test-utils'

const localVue = createLocalVue()
const pinia = createTestingPinia()
const tMock = {
  $t: text => text,
}

describe('Appointments', () => {
  let vuetify
  let wrapper

  beforeEach(() => {
    vuetify = new Vuetify()
    //@ts-ignore
    wrapper = mount(Appointments, {
      localVue,
      vuetify,
      pinia,
      mocks: tMock,
    })
  })
  afterEach(() => {
    wrapper.destroy()
  })

  it('should match the snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
