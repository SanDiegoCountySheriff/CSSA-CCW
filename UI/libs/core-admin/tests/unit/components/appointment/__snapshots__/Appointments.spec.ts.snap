// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appointments should match the snapshot 1`] = `
<div class="appointment-table">
  <div class="v-data-table v-data-table--has-top v-data-table--has-bottom theme--light">
    <header class="v-sheet theme--light v-toolbar v-toolbar--flat" style="height: 64px;">
      <div class="v-toolbar__content" style="height: 64px;">
        <div class="v-toolbar__title text-no-wrap pr-4" style="text-overflow: clip;">Appointments</div>
        <div class="spacer"></div>
        <div class="container appointment-table__header__container">
          <div class="row justify-end">
            <div class="col-md-6 col">
              <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line v-text-field--placeholder">
                <div class="v-input__prepend-outer">
                  <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-filter theme--light"></i></div>
                </div>
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-6" class="v-label theme--light" style="left: 0px; position: absolute;">Filter</label><input id="input-6" type="text"></div>
                  </div>
                </div>
              </div>
            </div>
            <div class="mr-1 col-md-4 col"><button type="button" class="v-btn v-btn--is-elevated v-btn--fab v-btn--has-bg v-btn--round v-btn--rounded theme--light elevation-2 v-size--x-small primary" raised="true"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-upload theme--light"></i></span></button> <input label="Upload New Appointments" type="file" accept=".csv, .xls, .xlsx" class="d-none"></div>
            <div class="col-md-1 col"><a href="/documents/ccw-appointment-schedule-template.csv" target="_blank" class="v-btn v-btn--is-elevated v-btn--fab v-btn--has-bg v-btn--round v-btn--rounded theme--light elevation-2 v-size--x-small primary" raised="" aria-label="Download the template file here" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-download theme--light"></i></span></a><span class="v-tooltip v-tooltip--top"><!----></span></div>
          </div>
        </div>
      </div>
    </header>
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="" class="text-start" style="width: 1px; min-width: 1px;"><span></span></th>
            <th role="columnheader" scope="col" aria-label="STATUS" class="text-start"><span>STATUS</span></th>
            <th role="columnheader" scope="col" aria-label="APPLICANT NAME: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>APPLICANT NAME</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="ORDER ID: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>ORDER ID</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="PAYMENT: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>PAYMENT</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="DATE: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>DATE</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="TIME: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>TIME</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span></span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          </tr>
        </thead>
        <tbody>
          <tr class="v-data-table__empty-wrapper">
            <td colspan="8">No data available</td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="v-data-footer">
      <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-23" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">5</div><input aria-label="Rows per page:" id="input-23" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="[object Object]">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-footer__pagination">â€“</div>
      <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="First page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-skip-backward theme--light"></i></span></button><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-minus theme--light"></i></span></button></div><span>1</span>
      <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-plus theme--light"></i></span></button><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Last page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-skip-forward theme--light"></i></span></button></div>
    </div>
  </div>
  <div class="v-snack v-snack--bottom v-snack--has-background">
    <transition-stub name="v-snack-transition">
      <div class="v-snack__wrapper v-sheet theme--dark" style="display: none;">
        <div role="status" aria-live="polite" class="v-snack__content">
          Invalid file type provided.

        </div>
        <div class="v-snack__action "><button type="button" class="v-btn v-btn--text theme--dark v-size--default red--text v-snack__btn"><span class="v-btn__content">
        Close
      </span></button></div>
      </div>
    </transition-stub>
  </div>
</div>
`;
